src/Database/LSMTree/Internal/Run/Reader.hs

```diff
@@ -0,0 +1,165 @@
+{-# LANGUAGE BangPatterns    #-}
+{-# LANGUAGE LambdaCase      #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-- | A run that is being read incrementally.
+--
+module Database.LSMTree.Internal.Run.Reader (
+    Reader
+  , new
+  , next
+  , Result (..)
+  , BlobRef (..)
+  , close
+  ) where
+
+import           Control.Exception (handleJust)
+import           Control.Monad (guard)
+import           Control.Monad.Primitive (RealWorld)
+import           Data.IORef
+import           Data.Maybe (fromMaybe)
+import           Data.Primitive.ByteArray (newPinnedByteArray,
+                     unsafeFreezeByteArray)
+import           Data.Primitive.PrimVar
+import           Data.Word (Word16, Word32)
+import           Database.LSMTree.Internal.BitMath (roundUpToPageSize)
+import           Database.LSMTree.Internal.BlobRef (BlobSpan (..))
+import           Database.LSMTree.Internal.Entry
+import qualified Database.LSMTree.Internal.RawBytes as RB
+import           Database.LSMTree.Internal.RawOverflowPage (RawOverflowPage,
+                     rawBytesToOverflowPages)
+import           Database.LSMTree.Internal.RawPage
+import qualified Database.LSMTree.Internal.Run as Run
+import           Database.LSMTree.Internal.Run.FsPaths
+import           Database.LSMTree.Internal.Serialise
+import qualified System.FS.API as FS
+import           System.FS.API (HasBufFS, HasFS)
+
+-- | Allows reading the k\/ops of a run incrementally, using its own read-only
+-- file handle and in-memory cache of the current disk page.
+--
+-- Creating a 'Reader' does not increase the run's reference count, so make sure
+-- the run remains open while using the 'Reader'.
+--
+-- New pages are loaded when trying to read their first entry.
+--
+-- TODO(optimise): Reuse page buffers using some kind of allocator. However,
+-- deciding how long a page needs to stay around is not trivial.
+data Reader fhandle = Reader {
+      readerCurrentPage    :: !(IORef RawPage)
+      -- | The index of the entry to be returned by the next call to 'next'.
+    , readerCurrentEntryNo :: !(PrimVar RealWorld Word16)
+      -- | Read mode file handle into the run's k\/ops file. We rely on it to
+      -- track the position of the next disk page to read, instead of keeping
+      -- a counter ourselves.
+    , readerKOpsHandle     :: !fhandle
```

Comment:
> > The RunReader has its own file handle to the k/ops file.
> 
> Ah yes because we need to re-open it in normal (not O_DIRECT) mode for the linear scan.

Let's add that in a comment here